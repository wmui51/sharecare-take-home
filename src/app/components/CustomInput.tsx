import { ChangeEvent, FocusEvent } from 'react';
import InputAdornment from '@mui/material/InputAdornment';
import InputLabel from '@mui/material/InputLabel';
import Input from '@mui/material/Input';
import EditOutlinedIcon from '@mui/icons-material/EditOutlined';
import { styled } from '@mui/material/styles';

import CustomFormControl from './CustomFormControl';

const CssInput = styled(Input)({
  '&.MuiInputBase-root': {
    '& .Mui-error': {
      backgroundColor: '#fef2f2',
    },
    '&::before': {
      borderColor: 'rgba(0, 0, 0, 0.12)',
    },
    '&::after': {
      borderColor: '#00bfa5',
    },
    '&:hover': {
      '&::after': {
        borderColor: '#00bfa5',
      },
    },
  },
});

const CssLabel = styled(InputLabel)({
  '&.Mui-focused': {
    color: '#00bfa5',
  },
});

export default function CustomInput({
  hasError,
  id,
  defaultValue,
  disabled,
  errorLabel,
  htmlFor,
  label,
  onBlur,
  onChange,
  onFocus,
  placeholder,
  type,
  value,
}: {
  hasError?: boolean,
  id: string,
  defaultValue?: string,
  disabled?: boolean,
  errorLabel?: string,
  htmlFor: string,
  label: string,
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void,
  onChange: (event: ChangeEvent<HTMLInputElement>) => void,
  onFocus?: (event: FocusEvent<HTMLInputElement>) => void,
  placeholder?: string,
  type?: string,
  value: string|number,
}) {
  const forceShrinkProp = errorLabel === "Required field." ? {shrink: true} : {};
  const customErrorBackground = hasError ? {backgroundColor: '#fef2f2'} : {};
  const customErrorAdornment = hasError ? {color: '#d32f2f'} : {};

  return (
    <CustomFormControl hasError={hasError} variant="standard" sx={customErrorBackground}>
      <CssLabel
        htmlFor={htmlFor}
        sx={{px: 2}}
        {...forceShrinkProp}
      >
        {hasError ? errorLabel : label}
      </CssLabel>
      <CssInput
        id={id}
        defaultValue={defaultValue}
        disabled={disabled}
        endAdornment={
          <InputAdornment position="end">
            <EditOutlinedIcon sx={customErrorAdornment} />
          </InputAdornment>
        }
        onBlur={onBlur}
        onChange={onChange}
        onFocus={onFocus}
        placeholder={placeholder}
        size="small"
        sx={{px: 2, py: 1}}
        type={type}
        value={value}
      />
    </CustomFormControl>
  );
}