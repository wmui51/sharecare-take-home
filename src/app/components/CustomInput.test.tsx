import { render } from '@testing-library/react';

import CustomInput from './CustomInput';

describe('<CustomInput />', () => {
  it('renders', () => {
    const { asFragment } = render(<CustomInput id="test" htmlFor="test" onChange={(event) => {}} label="label" value="test" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
