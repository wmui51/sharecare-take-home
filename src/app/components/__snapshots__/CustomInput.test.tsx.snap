// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomInput /> renders 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1m4d3n6-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="test"
    >
      label
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-c9goa0-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-1wd3yy0-MuiInputBase-input-MuiInput-input"
        id="test"
        type="text"
        value="test"
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-elo8k2-MuiInputAdornment-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="EditOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m14.06 9.02.92.92L5.92 19H5v-.92zM17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29m-3.6 3.19L3 17.25V21h3.75L17.81 9.94z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
